sunpy/map/mapbase.py

```diff
@@ -1452,58 +1455,43 @@ def submap(self, bottom_left, top_right=None):
             if not top_right:
                 if bottom_left.shape[0] != 2:
                     raise ValueError("If top_right is not specified bottom_left must have length two.")
                 else:
-                    lon, lat = self._get_lon_lat(bottom_left)
-                    top_right = u.Quantity([lon[1], lat[1]])
-                    bottom_left = u.Quantity([lon[0], lat[0]])
-            else:
-                bottom_left = u.Quantity(self._get_lon_lat(bottom_left))
-                top_right = u.Quantity(self._get_lon_lat(top_right))
-
-            top_left = u.Quantity([bottom_left[0], top_right[1]])
-            bottom_right = u.Quantity([top_right[0], bottom_left[1]])
-
-            corners = u.Quantity([bottom_left, bottom_right, top_left, top_right])
-            coord = SkyCoord(corners, frame=self.coordinate_frame)
-            pixel_corners = self.world_to_pixel(coord)
-
-            # Round the pixel values, we use floor+1 so that we always have at
-            # least one pixel width of data.
-            x_pixels = u.Quantity([np.min(pixel_corners.x), np.max(pixel_corners.x)]).to_value(u.pix)
-            x_pixels[0] = np.ceil(x_pixels[0])
-            x_pixels[1] = np.floor(x_pixels[1] + 1)
-            y_pixels = u.Quantity([np.min(pixel_corners.y), np.max(pixel_corners.y)]).to_value(u.pix)
-            y_pixels[0] = np.ceil(y_pixels[0])
-            y_pixels[1] = np.floor(y_pixels[1] + 1)
-
-        elif (isinstance(bottom_left, u.Quantity) and bottom_left.unit.is_equivalent(u.pix) and
-              isinstance(top_right, u.Quantity) and top_right.unit.is_equivalent(u.pix)):
-            x_pixels = u.Quantity([bottom_left[0], top_right[0]]).to_value(u.pix)
-            y_pixels = u.Quantity([top_right[1], bottom_left[1]]).to_value(u.pix)
+                    top_right = bottom_left[1]
+                    bottom_left = bottom_left[0]
 
-        else:
+            bottom_left = self.world_to_pixel(bottom_left)
+            top_right = self.world_to_pixel(top_right)
+
+        elif not (isinstance(bottom_left, u.Quantity) and bottom_left.unit.is_equivalent(u.pix) and
+                  isinstance(top_right, u.Quantity) and top_right.unit.is_equivalent(u.pix)):
             raise ValueError("Invalid input, bottom_left and top_right must either be SkyCoord or Quantity in pixels.")
 
+        x_pixels = u.Quantity([bottom_left[0], top_right[0]]).to_value(u.pix)
+        y_pixels = u.Quantity([top_right[1], bottom_left[1]]).to_value(u.pix)
         # Sort the pixel values so we always slice in the correct direction
         x_pixels.sort()
         y_pixels.sort()
 
+        # Round down the lower left pixel to the nearest integer
+        x_pixels[0] = np.round(x_pixels[0])
+        y_pixels[0] = np.round(y_pixels[0])
+        # Round the top right pixel to the nearest integer, then add 1 for array indexing
+        x_pixels[1] = np.round(x_pixels[1]) + 1
+        y_pixels[1] = np.round(y_pixels[1]) + 1
```

Comment:
Following on from mycomment in the docstring about including a pixel with any part, should this not always round down? Or am I mis-interpreting where we are indexing pixels from, i.e. are they referenced from the centre (FITS WCS) or bottom left (array indexing)?

If your "top_right" pixel coord is `(5.8, 5.8)`, it rounds up to 6 and then gets one added and ends up at 7, which is one pixel too large?

Reply:
This rounds up because when you're slicing an array, the "stop" part of the slice has to be one larger than the pixel number you want (e.g. to get pixel number 3 you would slice [3:4].

Reply:
I understand that as to why the +1 is there, but what I am trying to understand is the combo of the round up and the +1. 

So am I right in saying that if you have a 3x3 array and indexing from `(0, 0)` left you submap from `(0, 0)` to `(1.8, 1.8)` what's the output we expect? I think it should be a 2x2 array of the first 4 pixels right? i.e. the same as `[0:2]`

Well if you round up 1.8 and then add 1 you are doing `[0:3]` which gives you a 3x3 array out:
```python
>>> a                                                                                                                                                                                                                                                          
array([[4, 6, 2],
       [0, 4, 7],
       [1, 8, 8]])
>>> a[0:int(np.round(1.8) + 1), 0:int(np.round(1.8) + 1)]                                                                                                                                                                                                          
array([[4, 6, 2],
       [0, 4, 7],
       [1, 8, 8]])
```

If we are defining the centre of the pixels to be the index point (so our 3x3 array goes from `(-0.5, -0.5) -> (2.5, 2.5)`) then what we have is closer to correct, but I think we have issues with round down then (I would need to brush up on my WCS indexing).

---

(This is why we should use get item syntax for submap operations in pixel coordinates...)

Reply:
The 1.8 is rounded up because it has gone over the pixel border between pixels 1 and 2, that lies at a value of 1.5. Perhaps a picture helps:

<img width="419" alt="Screenshot 2020-05-11 at 10 21 26" src="https://user-images.githubusercontent.com/6197628/81545784-49fb1280-9371-11ea-9ce9-0f1db068f775.png">

Reply:
I wonder if using an implementation like this: https://github.com/spacetelescope/gwcs/blob/master/gwcs/utils.py#L52 which has been tested in a WCS framework would be a good idea?

I will think about this a little more, now I know we are in ints at centre of pixel mode.

Reply:
`floor(x + 0.5)` is the same as `round(x)`, so I the gwcs stuff is functionally doing the same. Edge cases (ie. 0.5) might be different though.

Reply:
> Edge cases (ie. 0.5) might be different though.

Yeah, definitely don't use `np.round()` for this code because it follows the convention of rounding to the closest even number.

Reply:
Woops, thanks for spotting that. As a note to myself, we want to make sure that for an edge (ie. 0.5) we include the pixel inside the rectangle but exclude the pixel outside.

Reply:
I've updated the rounding so that it correctly takes into account boundary (ie. n + 0.5) values.